parfive/config.py

```diff
@@ -0,0 +1,215 @@
+import os
+import sys
+import warnings
+from typing import Dict, Union, Optional
+
+try:
+    from typing import Literal  # Added in Python 3.8
+except ImportError:
+    from typing_extensions import Literal  # type: ignore
+
+from dataclasses import InitVar, field, asdict, dataclass
+
+import aiohttp
+
+import parfive
+from parfive.utils import ParfiveUserWarning
+
+__all__ = ["DownloaderConfig", "SessionConfig"]
+
+
+def _default_headers():
+    return {
+        "User-Agent": f"parfive/{parfive.__version__} aiohttp/{aiohttp.__version__} python/{sys.version[:5]}"
+    }
+
+
+@dataclass
+class SessionConfig:
```

Comment:
Maybe you are already aware of it but you could use [pydantic to capture settings form environment variables](https://pydantic-docs.helpmanual.io/usage/settings/). It also has support for validators, which can replace your `__post_init__` with much more targeted per field methods.
I think it will also help with readability. It will remove a lot of custom designed methods.


Here an example of how it could look like (to better understand this I suggest looking a bit into the docs for pydantic):

```python
from pydantic import BaseSettings, Field
from typing import Optional, Dict


class SessionConfig(BaseSettings):
    http_proxy: Optional[str] = Field(
        None,
        description="The URL of a proxy to use for HTTP requests.",
        env="HTTP_PROXY",
    )

    headers: Optional[Dict[str, str]] = Field(
        None,
        description=(
            "Headers to be passed to all requests made by this session. These headers"
            "are passed to the `aiohttp.ClientSession` along with"
            "``aiohttp_session_kwargs``."
            "The default value for headers is setting the user agent to a string with"
            "the version of parfive, aiohttp and Python."
            "To use aiohttp's default headers set this argument to an empty dictionary."
        ),
    )

    notebook: Optional[bool] = Field(
        None,
        description="""
     If `None` `tqdm` will automatically detect if it can draw rich IPython
     Notebook progress bars. If `False` or `True` notebook mode will be forced
     off or on.
     """,
    )
    aiohttp_session_kwargs: Dict = Field(
        {},
        description="""
    Any extra keyword arguments to be passed to `aiohttp.ClientSession`.
    Note that the `headers` keyword argument is handled separately, so should
    not be included in this dict.
    """,
    )

```

Reply:
Hmm, I did breifly consider using pydantic but decided the extra dep wasn't worth it, but I had forgotten it had env var parsing built in, that's probably enough to sway me (given pydantic itself has no real deps).

Reply:
Since we switched to it, we are really happy! it helps a lot on the long run. It pays back quite a lot.